### PR TITLE
[13571] Updating idme and logingov services to take advantage of JWK auto verification functionality

### PR DESCRIPTION
**Pull request by bosawt**
## Summary

- This PR updates functionality for id.me and login.gov services so that they automatically pull from each respective services public JWK endpoint and uses these JWKs properly and efficiently to verify the service responses from both of these entities. This should effectively not have an effect on visible production functionality, as these keys should be automatically pulled and verified. We should make sure that neither public endpoint is being spammed with every authentication though. A cached instance should be created once every production deploy (for each production instance)

## Related issue(s)
- department-of-veterans-affairs/vets-api#13571

## Testing done

- Authenticated with id.me loa3 user
- Authenticated with login.gov ial2 user

## What areas of the site does it impact?
Authentication

## Acceptance criteria

- [ ] Authenticate with id.me with Sign in Service and confirm it is working as expected
- [ ] Authenticate with login.gov with Sign in Service and confirm it is working as expected

### CODE REVIEW

File: lib/sign_in/idme/service.rb [hunk 1]
```diff
@@ -147,18 +165,18 @@ def jwe_decrypt(encrypted_jwe)
...
-          config.jwt_decode_public_key,
-          with_validation,
-          {
-            verify_expiration: with_validation,
-            algorithm: config.jwt_decode_algorithm
-          }
+          nil,
+          verify_expiration,
+          { verify_expiration:, algorithm: config.jwt_decode_algorithm, jwks: get_public_jwks }
         )&.first
```
Comment (dickdavis) [inline lib/sign_in/idme/service.rb:174]: ❓ I know you didn't touch this, but the use of the safe navigator here makes me wonder if it makes sense to handle a `nil` decoded JWT differently here. Maybe raise an `JWT::DecodeError` if the decoded JWT is `nil`?

```ruby
decoded_jwt = JWT.decode(
  encoded_jwt,
  nil,
  verify_expiration,
  { verify_expiration, algorithm: config.jwt_decode_algorithm, jwks: get_public_jwks }
)
raise JWT::DecodeError if decoded_jwt.blank?
```

Comment (dickdavis) [reply lib/sign_in/idme/service.rb:174]: [It doesn't look](https://github.com/jwt/ruby-jwt/blob/main/lib/jwt/decode.rb#L24) like `JWT.decode` can return `nil`, so I wonder why the safe navigation operator was introduced here. We could probably just call `.first` without using safe navigation, but if there is a chance that the decoded token could be `nil`, we should probably handle it.

**Review by joeniquette — COMMENTED** [review]
I tested with user 0 and 10 for idme, and Im getting a 400 error on the frontend. I just ran a db:migrate so I dont think thats an issue. Could be local setup problem. Here are the two lines in my rails log that may explain whats going on:

``` 
2023-08-22 11:50:07.020919 I [33265:puma srv tp 007] Rails -- [SignIn][Idme][Service] Token Success, code: 0116409ae05f4950b4dad138b2846a06, scope: http://idmanagement.gov/ns/assurance/loa/1/vets
2023-08-22 11:50:07.022477 E [33265:puma srv tp 007 sentry_logging.rb:57] Rails -- undefined method `access_token' for {:access_token=>"56e60834823c490ea0f0a52166846152", :token_type=>"bearer", :expires_in=>300, :scope=>"http://idmanagement.gov/ns/assurance/loa/1/vets", :refresh_token=>"e2a4451cf3da4436a2142914021ffa86", :refresh_expires_in=>604800}:Hash
```

**Comment by dickdavis** [comment]
> I tested with user 0 and 10 for idme, and Im getting a 400 error on the frontend. I just ran a db:migrate so I dont think thats an issue. Could be local setup problem. Here are the two lines in my rails log that may explain whats going on:
> 
> ```
> 2023-08-22 11:50:07.020919 I [33265:puma srv tp 007] Rails -- [SignIn][Idme][Service] Token Success, code: 0116409ae05f4950b4dad138b2846a06, scope: http://idmanagement.gov/ns/assurance/loa/1/vets
> 2023-08-22 11:50:07.022477 E [33265:puma srv tp 007 sentry_logging.rb:57] Rails -- undefined method `access_token' for {:access_token=>"56e60834823c490ea0f0a52166846152", :token_type=>"bearer", :expires_in=>300, :scope=>"http://idmanagement.gov/ns/assurance/loa/1/vets", :refresh_token=>"e2a4451cf3da4436a2142914021ffa86", :refresh_expires_in=>604800}:Hash
> ```

Looks like the change [here](https://github.com/department-of-veterans-affairs/vets-api/blob/7bccebbb7fcb3fca7e1c4962f3a5bb8b6693e233/app/controllers/v0/sign_in_controller.rb#L62) broke that. I think the `token` method on both of those services (`SignIn::Idme::Service` and `SignIn::Logingov::Service`) would need to change to return an `OpenStruct` instead of a `Hash` if we wanted to use method lookups. I'm generally against `OpenStructs` for performance reasons, though, so maybe it just needs to change back to a hash lookup.

**Comment by bosawt** [comment]
> > I tested with user 0 and 10 for idme, and Im getting a 400 error on the frontend. I just ran a db:migrate so I dont think thats an issue. Could be local setup problem. Here are the two lines in my rails log that may explain whats going on:
> > ```
> > 2023-08-22 11:50:07.020919 I [33265:puma srv tp 007] Rails -- [SignIn][Idme][Service] Token Success, code: 0116409ae05f4950b4dad138b2846a06, scope: http://idmanagement.gov/ns/assurance/loa/1/vets
> > 2023-08-22 11:50:07.022477 E [33265:puma srv tp 007 sentry_logging.rb:57] Rails -- undefined method `access_token' for {:access_token=>"56e60834823c490ea0f0a52166846152", :token_type=>"bearer", :expires_in=>300, :scope=>"http://idmanagement.gov/ns/assurance/loa/1/vets", :refresh_token=>"e2a4451cf3da4436a2142914021ffa86", :refresh_expires_in=>604800}:Hash
> > ```
> 
> Looks like the change [here](https://github.com/department-of-veterans-affairs/vets-api/blob/7bccebbb7fcb3fca7e1c4962f3a5bb8b6693e233/app/controllers/v0/sign_in_controller.rb#L62) broke that. I think the `token` method on both of those services (`SignIn::Idme::Service` and `SignIn::Logingov::Service`) would need to change to return an `OpenStruct` instead of a `Hash` if we wanted to use method lookups. I'm generally against `OpenStructs` for performance reasons, though, so maybe it just needs to change back to a hash lookup.

Oooh I didn't know about `OpenStruct` not being performant. Given that this is just creating a single one rather than an array or anything like that, this surely won't cause any perf issue. But we're just using this as a hash so prob can just make both services return hashes

File: lib/sign_in/idme/service.rb (hunk 1 above)
Comment (bosawt) [reply lib/sign_in/idme/service.rb:174]: I think the error handling should cover any instances that would result in a `nil` jwt so I'll remove the safe navigator in these

**Review by joeniquette — APPROVED** [review]
both idme and login worked, vets gov user 0

**Pull request merged**